packages/store/ts/codegen/record.ts

```diff
@@ -115,69 +120,95 @@ function renderDecodeFunction({ structName, fields, staticFields, dynamicFields
     _acc += field.staticByteLength;
   }
 
+  let result = "";
+
+  if (staticFields.length > 0) {
+    result += `
+    /**
+     * Decode the tightly packed blob of static data using this table's field layout
+     * Undefined behaviour for invalid blobs
+     */
+    function decodeStatic(bytes memory _blob) internal pure returns (${renderArguments(
+      staticFields.map(({ name, typeWithLocation }) => `${typeWithLocation} ${name}`)
+    )}) {
+      ${renderList(
+        staticFields,
+        (field, index) => `
+        ${field.name} = ${renderDecodeValueType(field, staticOffsets[index])};
+        `
+      )}
+    }
+    `;
+  }
+
   if (dynamicFields.length > 0) {
-    const totalStaticLength = staticFields.reduce((acc, { staticByteLength }) => acc + staticByteLength, 0);
-    // decode static (optionally) and dynamic data
-    return `
-      /**
-       * Decode the tightly packed blob using this table's field layout.
-       * Undefined behaviour for invalid blobs.
-       */
-      function decode(bytes memory _blob) internal pure returns (${renderedDecodedRecord}) {
-        // ${totalStaticLength} is the total byte length of static data
-        PackedCounter _encodedLengths = PackedCounter.wrap(Bytes.slice32(_blob, ${totalStaticLength})); 
-
-        ${renderList(
-          staticFields,
-          (field, index) => `
-            ${fieldNamePrefix}${field.name} = ${renderDecodeValueType(field, staticOffsets[index])};
-          `
-        )}
-        // Store trims the blob if dynamic fields are all empty
-        if (_blob.length > ${totalStaticLength}) {
-          ${renderList(
-            dynamicFields,
-            // unchecked is only dangerous if _encodedLengths (and _blob) is invalid,
-            // but it's assumed to be valid, and this function is meant to be mostly used internally
-            (field, index) => {
-              if (index === 0) {
-                return `
-                  // skip static data length + dynamic lengths word
-                  uint256 _start = ${totalStaticLength + 32};
-                  uint256 _end;
-                  unchecked {
-                    _end = ${totalStaticLength + 32} + _encodedLengths.atIndex(${index});
-                  }
-                  ${fieldNamePrefix}${field.name} = ${renderDecodeDynamicFieldPartial(field)};
-                `;
-              } else {
-                return `
-                  _start = _end;
-                  unchecked {
-                    _end += _encodedLengths.atIndex(${index});
-                  }
-                  ${fieldNamePrefix}${field.name} = ${renderDecodeDynamicFieldPartial(field)};
-                `;
+    result += `
+    /**
```

Comment:
nit: can we indent this one level (and all the backticks below) for readability?

Reply:
sure, I went with this style for all renderers so just sticking to it here. If we wanna change it I'd rather make a separate PR and change it everywhere

Reply:
we already did this: https://github.com/latticexyz/mud/pull/1474